src/GafferScene/CopyPrimitiveVariables.cpp

```diff
@@ -0,0 +1,208 @@
+	if( !copyFrom.empty() )
+	{
+		/// \todo When we can use `std::optional` from C++17, `emplace()`
+		/// will return a reference, allowing us to call
+		/// `stringToPath( copyFrom, copyFromPath.emplace() )`.
+		copyFromPath.emplace();
+		ScenePlug::stringToPath( copyFrom, *copyFromPath );
+	}
+
+	/// \todo Consider the performance implications of calling `exists()` here.
+	/// Perhaps we need to introduce an element of caching via
+	/// a `ScenePlug::existsPlug()`, the computation of which can depend on the
+	/// (most likely cached) computation of the ancestor location.
+	if( !SceneAlgo::exists( inPlugs()->getChild<ScenePlug>( 1 ), copyFromPath ? *copyFromPath : path ) )
+	{
+		h = inPlug()->objectPlug()->hash();
+		return;
+	}
+
+	if( copyFromPath )
+	{
+		h.append( inPlugs()->getChild<ScenePlug>( 1 )->objectHash( *copyFromPath ) );
+	}
+	else
+	{
+		inPlugs()->getChild<ScenePlug>( 1 )->objectPlug()->hash( h );
+	}
+}
+
+IECore::ConstObjectPtr CopyPrimitiveVariables::computeProcessedObject( const ScenePath &path, const Gaffer::Context *context, const IECore::Object *inputObject ) const
+{
+	auto primitive = runTimeCast<const Primitive>( inputObject );
+	if( !primitive )
+	{
+		return inputObject;
+	}
+
+	const string primitiveVariables = primitiveVariablesPlug()->getValue();
+	if( primitiveVariables.empty() )
+	{
+		return inputObject;
+	}
+
+	boost::optional<ScenePath> copyFromPath;
+	const string copyFrom = copyFromPlug()->getValue();
+	if( !copyFrom.empty() )
+	{
+		/// \todo When we can use `std::optional` from C++17, `emplace()`
+		/// will return a reference, allowing us to call
+		/// `stringToPath( copyFrom, copyFromPath.emplace() )`.
+		copyFromPath.emplace();
+		ScenePlug::stringToPath( copyFrom, *copyFromPath );
+	}
+
+	if( !SceneAlgo::exists( inPlugs()->getChild<ScenePlug>( 1 ), copyFromPath ? *copyFromPath : path ) )
+	{
+		return inputObject;
+	}
+
+	ConstObjectPtr sourceObject;
+	if( copyFromPath )
+	{
+		sourceObject = inPlugs()->getChild<ScenePlug>( 1 )->object( *copyFromPath );
+	}
+	else
+	{
+		sourceObject = inPlugs()->getChild<ScenePlug>( 1 )->objectPlug()->getValue();
+	}
+
+	auto sourcePrimitive = runTimeCast<const Primitive>( sourceObject.get() );
+	if( !sourcePrimitive )
+	{
+		return inputObject;
+	}
+
+	PrimitivePtr result = primitive->copy();
+	for( auto &variable : sourcePrimitive->variables )
+	{
+		if( !StringAlgo::matchMultiple( variable.first, primitiveVariables ) )
+		{
+			continue;
+		}
+		if( !result->isPrimitiveVariableValid( variable.second ) )
+		{
+			string destinationPath; ScenePlug::pathToString( path, destinationPath );
+			const string &sourcePath = copyFrom.size() ? copyFrom : destinationPath;
+			throw IECore::Exception( boost::str(
+				boost::format( "Cannot copy \"%1%\" from \"%2%\" to \"%3%\" because source and destination primitives have different topology" )
+					% variable.first % destinationPath % sourcePath
+			) );
+		}
+		result->variables[variable.first] = variable.second;
```

Comment:
I thought there was some cost to the `result->variables` lookup and it was better to do that outside the loop.

Reply:
Can you elaborate? As far as I know it's just a pointer offset to get at the member : https://github.com/ImageEngine/cortex/blob/master/include/IECoreScene/Primitive.h#L68.

Reply:
Its just something I've gotten in the habit of... maybe it matters more in python?

Reply:
I don't understand; in Python we don't even expose the `Primitive::variables` member directly, because you get at variables through dictionary access on the primitive object itself (`prim["P"]` for instance). Are we talking about the same thing?